clash-prelude/src/Clash/Class/HasDomain/CodeGen.hs

```diff
@@ -0,0 +1,36 @@
+import Data.List (intercalate)
+
+data WhichDomain = HasDomain | HasDomainL | HasDomainR
+
+commas :: [String] -> String
+commas = intercalate ", "
+
+domainEq :: String -> String -> String
+domainEq a b = a ++ " `DomainEq` " ++ b
+
+inst :: WhichDomain -> Integer -> IO ()
+inst d n = do
+  case d of
+    HasDomain -> do
+      putStrLn $ "instance (" ++ constraints ++ ") => HasDomain (" ++ tup ++ ") where"
+      putStrLn $ "  type GetDomain (" ++ tup ++ ") = GetDomain " ++ head vars
+      putStrLn ""
```

Comment:
I feel like this is something you can do fairly easily with template Haskell without having to add ~700 lines of generated code to git.

Reply:
Hmmyeah, I tried it with template Haskell, but it's _so slow_. I do mention this in `HasDomain{N,L,R}`:

```
-- Tuple instances. Generated with 'HasDomain/CodeGen.hs". Purposely generated
-- with a script to prevent long Template Haskell compile times.
```

but that comment should probably be added to `CodeGen.hs` too.

Reply:
I ran some tests this morning: reducing the number of tuple instances that are generated by Template Haskell from 63 to 7 (in Bitpack/Bundle) halves the compile time. So I'll probably add a CPP flag developers can override in their `cabal.project.local` indicating how many instances should be generated..